### DISCUSSION TITLE
A few change ideas on getline

### DISCUSSION THREAD

**Discussion by kuaqe** [msteveb/jimtcl#232] [Ideas]
Hello, I'm learning jimtcl and I'm doing a sqlite backed terminal app. The users can add new db entries but they should also be able to edit existing ones. So I need a way to feed the initial line to getline. I modified the readline extension[0] for the purpose as it felt easier. But it would be nice to have the same mechanism with `history getline`. If you think it's a good idea I can give a shot at implementing it. Maybe you can also let me know what you think about adding the word motion keys M-b M-f and M-w (back forward and delete word). Thank you!

Edit: Another thing, it would also be useful to know if the user pressed C-d or just Enter on empty input. Right now there is no way of distinguishing between the two.

[0]
```
static const char *deftext;

static int set_deftext()
{
	rl_insert_text(deftext);
	return 0;
}

static int JimRlReadlineCommand(Jim_Interp *interp, int argc, Jim_Obj *const *argv)
{
    char *line;

    if (argc < 2 || argc > 3) {
        Jim_WrongNumArgs(interp, 1, argv, "prompt ?preGiven");
        return JIM_ERR;
    }
    if (argc == 3) {
	    deftext = Jim_String(argv[2]);
	    rl_startup_hook = set_deftext;
    }
    line = readline(Jim_String(argv[1]));
    rl_startup_hook = NULL;
    if (!line) {
        return JIM_EXIT;
    }
    Jim_SetResult(interp, Jim_NewStringObj(interp, line, -1));
    return JIM_OK;
}
```

**Comment by msteveb** [comment]
The ability to provide an initial string to edit does seem useful. The underlying support in linenoise would need to come through here though: https://github.com/msteveb/linenoise

Then we could pull it into jimtcl and add it to the history extension.

Regarding extra key bindings, again they would need to come through linenoise. We already have ctrl-w for delete word to left but not move word left or right.

**Comment by msteveb** [comment]
I did like the idea of extra key bindings. See here - https://github.com/msteveb/linenoise/commit/6cd2bcba97950bb2cafe460eb5558f41c64e081a

And here for the ability to specify an initial value - https://github.com/msteveb/linenoise/commit/d26a12bc449102c9b50a7623afe5559747eb291e

Not yet integrated into JimTcl and history though

**Comment by kuaqe** [comment]
Thank you @msteveb!

**Reply by msteveb** [reply]
Soon(ish) I will import this version into jimtcl, but if you want integration with history you will probably need do it yourself and submit a pull request.

**Comment by msteveb** [comment]
This has now been integrated
